src/File/RenameUpload.php

```diff
@@ -27,207 +28,76 @@
 use function unlink;
 
 use const DIRECTORY_SEPARATOR;
+use const PATHINFO_DIRNAME;
+use const PATHINFO_EXTENSION;
+use const PATHINFO_FILENAME;
 use const UPLOAD_ERR_OK;
 
 /**
+ * @psalm-type UploadedFile = array{
+ *     name:string,
+ *     tmp_name: string
+ * }
  * @psalm-type Options = array{
- *     target: string|null,
- *     use_upload_name: bool,
- *     use_upload_extension: bool,
- *     overwrite: bool,
- *     randomize: bool,
- *     stream_factory: StreamFactoryInterface|null,
- *     upload_file_factory: UploadedFileFactoryInterface|null,
- *     ...
+ *     target?: string,
+ *     use_upload_name?: bool,
+ *     use_upload_extension?: bool,
+ *     overwrite?: bool,
+ *     randomize?: bool,
+ *     stream_factory?: StreamFactoryInterface,
+ *     upload_file_factory?: UploadedFileFactoryInterface,
  * }
- * @template TOptions of Options
- * @extends AbstractFilter<TOptions>
+ * @implements FilterInterface<mixed>
+ * @final
```

Comment:
Used this `@final` instead of making the class final because otherwise we could not overwrite the `move_uploaded_file` function. 

Or do you prefer to pass a dependency via option that could change the behavior via DependencyInjection? And should the option be documented as well?

Reply:
I added the deprecated aswell for these methods here #224

Reply:
We should be able to make this final without any problems… and IMO it should be `@implements FilterInterface<string>`, and operate on a single upload at a time. If it allows multiple uploads, then we should change that.

Reply:
I think we need to create a factory for this filter and register it with the "FilterManager". The constructor signature should change to something like:

`__construct(StreamFactoryInterface, UploadedFileFactoryInterface, Closure, Options)`

Where the Closure wraps `move_uploaded_file`, so we can easily inject something else in tests. We can make this closure within the factory instead of registering it as a service.

It _may_ be necessary to make PSR StreamFactory and PSR UploadFactory nullable for compat with MVC.